File: DexBar/Sources/Managers/GlucoseMonitor.swift
import AppKit
import DexBarCore
import Foundation
import Observation
import SwiftUI

enum MenuBarStyle: String, CaseIterable {
    case full      = "Value & Arrow"
    case compact   = "Compact"
    case valueOnly = "Value Only"
    case arrowOnly = "Arrow Only"
}

@MainActor
@Observable
final class GlucoseMonitor {
    // Current state
    var currentReading: GlucoseReading?
    var recentReadings: [GlucoseReading] = []   // newest first, up to 90 days
    var selectedTimeRange: TimeRange = .threeHours
    var selectedStatsRange: StatsTimeRange = .sevenDays
    var state: MonitorState = .idle
    var lastUpdated: Date?

    var chartReadings: [GlucoseReading] {
        let cutoff = Date().addingTimeInterval(-selectedTimeRange.interval)
        return recentReadings.filter { $0.date >= cutoff }
    }

    private var statsReadings: [GlucoseReading] {
        let cutoff = Date().addingTimeInterval(-selectedStatsRange.interval)
        return recentReadings.filter { $0.date >= cutoff }
    }

    /// Actual days of data available for the selected stats range.
    var statsDataSpanDays: Double {
        guard let oldest = statsReadings.last?.date else { return 0 }
        return Date().timeIntervalSince(oldest) / 86400
    }

    var tirStats: TiRStats {
        let readings = statsReadings
        let lowCutoff  = alertLowThresholdMgdL
        let highCutoff = alertHighThresholdMgdL
        let low  = readings.filter { Double($0.value) < lowCutoff  }.count
        let high = readings.filter { Double($0.value) > highCutoff }.count
        return TiRStats(lowCount: low, inRangeCount: readings.count - low - high, highCount: high, total: readings.count)
    }

    /// Glucose Management Indicator — estimated HbA1c % from mean glucose.
    /// Formula: GMI = 3.31 + 0.02392 × mean_mg_dL
    var gmi: Double? {
        let readings = statsReadings
        guard !readings.isEmpty else { return nil }
        let mean = Double(readings.reduce(0) { $0 + $1.value }) / Double(readings.count)
        return 3.31 + 0.02392 * mean
    }

    // Settings (persisted via AppStorage in SettingsView; mirrored here)
    var unit: GlucoseUnit = .mgdL
    var refreshInterval: TimeInterval = 5 * 60

    // Alert settings
    var alertUrgentHighEnabled: Bool = true
    var alertUrgentHighThresholdMgdL: Double = 250
    var alertHighEnabled: Bool = true
    var alertHighThresholdMgdL: Double = 180
    var alertLowEnabled: Bool = true
    var alertLowThresholdMgdL: Double = 70
    var alertUrgentLowEnabled: Bool = true
    var alertUrgentLowThresholdMgdL: Double = 55
    var alertRisingFastEnabled: Bool = true
    var alertDroppingFastEnabled: Bool = true
    var alertStaleDataEnabled: Bool = true
    var alertCriticalEnabled: Bool = UserDefaults.standard.object(forKey: "alertCriticalEnabled") as? Bool ?? false {
        didSet { UserDefaults.standard.set(alertCriticalEnabled, forKey: "alertCriticalEnabled") }
    }
    static let staleThreshold: TimeInterval = 20 * 60

    var isStale: Bool {
        guard let reading = currentReading else { return false }
        return Date().timeIntervalSince(reading.date) > Self.staleThreshold
    }

    // Zone colors (persisted in UserDefaults as hex strings)
    var colorUrgentLow: Color = Color(hex: UserDefaults.standard.string(forKey: "colorUrgentLow") ?? "") ?? Color(red: 0.85, green: 0.1, blue: 0.1) {
        didSet { if let h = colorUrgentLow.toHex() { UserDefaults.standard.set(h, forKey: "colorUrgentLow") } }
    }
    var colorLow: Color = Color(hex: UserDefaults.standard.string(forKey: "colorLow") ?? "") ?? .orange {
        didSet { if let h = colorLow.toHex() { UserDefaults.standard.set(h, forKey: "colorLow") } }
    }
    var colorInRange: Color = Color(hex: UserDefaults.standard.string(forKey: "colorInRange") ?? "") ?? .green {
        didSet { if let h = colorInRange.toHex() { UserDefaults.standard.set(h, forKey: "colorInRange") } }
    }
    var colorHigh: Color = Color(hex: UserDefaults.standard.string(forKey: "colorHigh") ?? "") ?? .yellow {
        didSet { if let h = colorHigh.toHex() { UserDefaults.standard.set(h, forKey: "colorHigh") } }
    }
    var colorUrgentHigh: Color = Color(hex: UserDefaults.standard.string(forKey: "colorUrgentHigh") ?? "") ?? Color(red: 0.85, green: 0.1, blue: 0.1) {
        didSet { if let h = colorUrgentHigh.toHex() { UserDefaults.standard.set(h, forKey: "colorUrgentHigh") } }
    }
    var coloredMenuBar: Bool = UserDefaults.standard.bool(forKey: "coloredMenuBar") {
        didSet { UserDefaults.standard.set(coloredMenuBar, forKey: "coloredMenuBar") }
    }
    var menuBarStyle: MenuBarStyle = MenuBarStyle(rawValue: UserDefaults.standard.string(forKey: "menuBarStyle") ?? "") ?? .full {
        didSet { UserDefaults.standard.set(menuBarStyle.rawValue, forKey: "menuBarStyle") }
    }
    var showDelta: Bool = UserDefaults.standard.object(forKey: "showDelta") as? Bool ?? true {
        didSet { UserDefaults.standard.set(showDelta, forKey: "showDelta") }
    }

    var glucoseDelta: Int? {
        guard recentReadings.count >= 2 else { return nil }
        return recentReadings[0].value - recentReadings[1].value
    }

    func formattedDelta(unit: GlucoseUnit) -> String? {
        guard let delta = glucoseDelta else { return nil }
        switch unit {
        case .mgdL:
            return delta >= 0 ? "+\(delta)" : "\(delta)"
        case .mmolL:
            let dMmol = Double(delta) / 18.0
            return dMmol >= 0 ? String(format: "+%.1f", dMmol) : String(format: "%.1f", dMmol)
        }
    }

    var readingColor: Color {
        guard let reading = currentReading else { return .primary }
        let v = Double(reading.value)
        if v < alertUrgentLowThresholdMgdL  { return colorUrgentLow  }
        if v < alertLowThresholdMgdL         { return colorLow         }
        if v > alertUrgentHighThresholdMgdL { return colorUrgentHigh }
        if v > alertHighThresholdMgdL        { return colorHigh        }
        return colorInRange
    }

    private var service: DexcomService?
    private var timer: Timer?
    var nextRefreshDate: Date?
    private var isStarting = false

    private static let readingsURL: URL? = {
        FileManager.default.urls(for: .applicationSupportDirectory, in: .userDomainMask).first?
            .appendingPathComponent("DexBar/readings.json")
    }()

    private func saveReadings() {
        guard let url = Self.readingsURL else { return }
        let dir = url.deletingLastPathComponent()
        try? FileManager.default.createDirectory(at: dir, withIntermediateDirectories: true)
        let data = try? JSONEncoder().encode(recentReadings)
        try? data?.write(to: url, options: .atomic)
    }

    private func loadPersistedReadings() {
        guard let url = Self.readingsURL,
              let data = try? Data(contentsOf: url),
              let readings = try? JSONDecoder().decode([GlucoseReading].self, from: data) else { return }
        recentReadings = readings
    }

    init() {
        loadPersistedReadings()
        Task { @MainActor in
            await autoConnectIfNeeded()
        }
        NSWorkspace.shared.notificationCenter.addObserver(
            forName: NSWorkspace.didWakeNotification,
            object: nil,
            queue: .main
        ) { [weak self] _ in
            guard let self else { return }
            Task { @MainActor in await self.handleSystemWake() }
        }
    }

    private func autoConnectIfNeeded() async {
        let username = UserDefaults.standard.string(forKey: "dexcomUsername") ?? ""
        let regionRaw = UserDefaults.standard.string(forKey: "dexcomRegion") ?? DexcomRegion.us.rawValue
        guard !username.isEmpty,
              let password = try? KeychainService.load(key: "password"),
              !password.isEmpty else { return }
        let region = DexcomRegion(rawValue: regionRaw) ?? .us
        await start(username: username, password: password, region: region)
    }

    // MARK: - Lifecycle

    func start(username: String, password: String, region: DexcomRegion) async {
        guard !isStarting else { return }
        isStarting = true
        defer { isStarting = false }
        service = DexcomService(region: region)
        state = .loading
        do {
            try await service?.authenticate(username: username, password: password)
        } catch {
            state = .error(error.localizedDescription)
            return
        }
        await refresh(initialLoad: true)
        // Timer is scheduled inside refresh() once the first reading is obtained
    }

    func stop() {
        timer?.invalidate()
        timer = nil
        nextRefreshDate = nil
        service = nil
        state = .idle
    }

    func refreshNow() async {
        await refresh(initialLoad: false)
    }

    func updateRefreshInterval(_ interval: TimeInterval) {
        refreshInterval = interval
        // Reschedule based on last reading so the window stays aligned
        if timer != nil {
            scheduleTimer(after: currentReading?.date)
        }
    }

    // MARK: - Private

    /// Schedule the next auto-refresh at `lastReadingDate + refreshInterval`.
    /// If that time is already past (or no reading yet), waits at least 30 s to avoid hammering the API.
    private func scheduleTimer(after lastReadingDate: Date? = nil) {
        timer?.invalidate()
        let fireDate: Date
        if let last = lastReadingDate {
            let candidate = last.addingTimeInterval(refreshInterval)
            fireDate = max(candidate, Date().addingTimeInterval(30))
        } else {
            fireDate = Date().addingTimeInterval(refreshInterval)
        }
        nextRefreshDate = fireDate
        timer = Timer(fire: fireDate, interval: 0, repeats: false) { [weak self] _ in
            guard let self else { return }
            Task { @MainActor in
                await self.refresh()
            }
        }
        RunLoop.main.add(timer!, forMode: .common)
    }

    private func refresh(initialLoad: Bool = false) async {
        guard let service else { return }
        state = .loading
        let maxCount = initialLoad ? 288 : 2
        do {
            let newReadings = try await service.getLatestReadings(maxCount: maxCount)
            let reading = newReadings[0]
            currentReading = reading
            // Merge new readings into history deduplicating by date, cap at 288
            let existingDates = Set(recentReadings.map { $0.date })
            let toAdd = newReadings.filter { !existingDates.contains($0.date) }
            let merged = (toAdd + recentReadings).sorted { $0.date > $1.date }
            recentReadings = Array(merged.prefix(25920))  // 90 days × 288 readings/day
            lastUpdated = Date()
            state = .connected
            await evaluateAlerts(reading: reading)
            await evaluateStaleAlert(reading: reading)
            saveReadings()
            scheduleTimer(after: reading.date)
        } catch DexcomError.sessionExpired, DexcomError.invalidCredentials {
            await reAuthenticateIfPossible()
        } catch DexcomError.serverError(let code) where code == 429 {
            state = .error("Rate limited by Dexcom — will retry soon")
            scheduleTimer()
        } catch {
            state = .error(error.localizedDescription)
            scheduleTimer(after: currentReading?.date)
        }
    }

    private func handleSystemWake() async {
        // Wait briefly for the network to reconnect before attempting a refresh.
        try? await Task.sleep(for: .seconds(3))
        guard service != nil else {
            await autoConnectIfNeeded()
            return
        }
        await refresh()
    }

    private func reAuthenticateIfPossible() async {
        let username = UserDefaults.standard.string(forKey: "dexcomUsername") ?? ""
        let regionRaw = UserDefaults.standard.string(forKey: "dexcomRegion") ?? DexcomRegion.us.rawValue
        guard !username.isEmpty,
              let password = try? KeychainService.load(key: "password"),
              !password.isEmpty else {
            state = .error("Session expired — reconnect in Settings")
            if let svc = service { await svc.clearSession() }
            scheduleTimer(after: currentReading?.date)
            return
        }
        let region = DexcomRegion(rawValue: regionRaw) ?? .us
        await start(username: username, password: password, region: region)
    }

    private func evaluateStaleAlert(reading: GlucoseReading) async {
        guard alertStaleDataEnabled else { return }
        let age = Date().timeIntervalSince(reading.date)
        guard age > Self.staleThreshold else { return }
        let minutes = Int(age / 60)
        await NotificationManager.shared.send(
            type: .staleData,
            title: "No New Readings",
            body: "Last reading was \(minutes) minutes ago. Check your sensor."
        )
    }

    private func evaluateAlerts(reading: GlucoseReading) async {
        let nm = NotificationManager.shared
        let displayVal = reading.displayValue(unit: unit)
        let unitStr = unit.rawValue
        let v = Double(reading.value)

        if alertUrgentHighEnabled, v > alertUrgentHighThresholdMgdL {
            await nm.send(type: .urgentHigh, title: "Urgent High Blood Sugar",
                body: "\(displayVal) \(unitStr) — urgently above your high threshold",
                isCritical: alertCriticalEnabled)
        } else if alertHighEnabled, v > alertHighThresholdMgdL {
            await nm.send(type: .high, title: "High Blood Sugar",
                body: "\(displayVal) \(unitStr) — above your high alert threshold")
        }

        if alertUrgentLowEnabled, v < alertUrgentLowThresholdMgdL {
            await nm.send(type: .urgentLow, title: "Urgent Low Blood Sugar",
                body: "\(displayVal) \(unitStr) — urgently below your low threshold",
                isCritical: alertCriticalEnabled)
        } else if alertLowEnabled, v < alertLowThresholdMgdL {
            await nm.send(type: .low, title: "Low Blood Sugar",
                body: "\(displayVal) \(unitStr) — below your low alert threshold")
        }

        if alertRisingFastEnabled, reading.trend.isRisingFast {
            await nm.send(type: .risingFast, title: "Blood Sugar Rising Fast",
                body: "\(displayVal) \(unitStr) and \(reading.trend.description)")
        }
        if alertDroppingFastEnabled, reading.trend.isDroppingFast {
            await nm.send(type: .droppingFast, title: "Blood Sugar Dropping Fast",
                body: "\(displayVal) \(unitStr) and \(reading.trend.description)")
        }
    }
}

File: DexBarWindows/Managers/GlucoseMonitor.cs
using System.Text.Json;
using System.Text.Json.Serialization;
using System.Threading;
using DexBarWindows.Models;
using DexBarWindows.Services;

namespace DexBarWindows.Managers;

/// <summary>
/// Manages glucose polling, state, alerts, and persisted readings.
/// Thread-safe state changes are marshalled back to the UI thread via the
/// SynchronizationContext captured at construction time.
/// </summary>
public class GlucoseMonitor : IDisposable
{
    // -------------------------------------------------------------------------
    // Constants
    // -------------------------------------------------------------------------

    private const int MaxReadings = 25_920;
    private const int InitialFetchCount = 288;
    private const int SubsequentFetchCount = 2;
    private const int MaxAuthRetries = 3;
    private static readonly TimeSpan StaleThreshold = TimeSpan.FromMinutes(20);
    private static readonly TimeSpan AlertCooldown = TimeSpan.FromMinutes(15);

    private static readonly int[] RetryDelaysMs = { 3_000, 5_000, 10_000 };

    private static readonly string ReadingsPath = Path.Combine(
        Environment.GetFolderPath(Environment.SpecialFolder.ApplicationData),
        "DexBar",
        "readings.json");

    private static readonly JsonSerializerOptions JsonOptions = new()
    {
        WriteIndented = false,
        Converters = { new JsonStringEnumConverter() }
    };

    // -------------------------------------------------------------------------
    // State
    // -------------------------------------------------------------------------

    private readonly SynchronizationContext _syncContext;
    private DexcomService? _service;
    private System.Threading.Timer? _timer;
    private string? _username;
    private bool _disposed;

    // Serialization lock: ensures only one poll runs at a time and protects
    // _recentReadings mutations.
    private readonly SemaphoreSlim _pollLock = new(1, 1);

    private readonly Dictionary<string, DateTime> _alertCooldowns = new();

    // -------------------------------------------------------------------------
    // Observable state (always accessed on UI thread via _syncContext)
    // -------------------------------------------------------------------------

    /// <summary>The most recent glucose reading, or null if none available.</summary>
    public GlucoseReading? CurrentReading { get; private set; }

    /// <summary>All readings, newest first, capped at 25,920 entries.</summary>
    public List<GlucoseReading> RecentReadings { get; private set; } = [];

    /// <summary>Current monitor state.</summary>
    public MonitorState State { get; private set; } = new MonitorState.Idle();

    /// <summary>When state was last updated.</summary>
    public DateTime? LastUpdated { get; private set; }

    /// <summary>When the next automatic refresh is scheduled.</summary>
    public DateTime? NextRefreshDate { get; private set; }

    /// <summary>Loaded app settings.</summary>
    public AppSettings Settings { get; private set; }

    // -------------------------------------------------------------------------
    // Callbacks
    // -------------------------------------------------------------------------

    /// <summary>Fired on any state change (always on UI thread).</summary>
    public Action? OnUpdate { get; set; }

    /// <summary>Fired when an alert condition is triggered (title, message).</summary>
    public Action<string, string>? OnAlert { get; set; }

    // -------------------------------------------------------------------------
    // Computed properties
    // -------------------------------------------------------------------------

    /// <summary>
    /// Delta between the two most recent readings (newest - second-newest), in mg/dL.
    /// Null if fewer than 2 readings.
    /// </summary>
    public int? GlucoseDelta =>
        RecentReadings.Count >= 2
            ? RecentReadings[0].Value - RecentReadings[1].Value
            : null;

    /// <summary>
    /// Formatted delta string such as "+3" or "-0.2", respecting the configured unit.
    /// Returns null if fewer than 2 readings.
    /// </summary>
    public string? FormattedDelta(GlucoseUnit unit)
    {
        var delta = GlucoseDelta;
        if (delta is null) return null;

        if (unit == GlucoseUnit.MmolL)
        {
            var mmol = delta.Value / 18.0;
            var sign = mmol >= 0 ? "+" : "";
            return $"{sign}{mmol:F1}";
        }
        else
        {
            var sign = delta.Value >= 0 ? "+" : "";
            return $"{sign}{delta.Value}";
        }
    }

    /// <summary>True if the current reading is older than 20 minutes (or there is none).</summary>
    public bool IsStale =>
        CurrentReading is null ||
        DateTime.UtcNow - CurrentReading.Date > StaleThreshold;

    /// <summary>
    /// Time-in-Range statistics computed over the configured StatsTimeRange window.
    /// </summary>
    public TiRStats TirStats
    {
        get
        {
            var cutoff = DateTime.UtcNow - Settings.StatsTimeRange.Interval();
            var window = RecentReadings.Where(r => r.Date >= cutoff).ToList();

            int low = 0, inRange = 0, high = 0;
            foreach (var r in window)
            {
                if (r.Value <= Settings.AlertLowThresholdMgdL)
                    low++;
                else if (r.Value < Settings.AlertHighThresholdMgdL)
                    inRange++;
                else
                    high++;
            }

            return new TiRStats
            {
                LowCount = low,
                InRangeCount = inRange,
                HighCount = high
            };
        }
    }

    /// <summary>
    /// Glucose Management Indicator (GMI) calculated from mean glucose in the stats window.
    /// Returns null if there are no readings in the window.
    /// </summary>
    public double? Gmi
    {
        get
        {
            var cutoff = DateTime.UtcNow - Settings.StatsTimeRange.Interval();
            var window = RecentReadings.Where(r => r.Date >= cutoff).ToList();
            if (window.Count == 0) return null;

            var mean = window.Average(r => r.Value);
            return 3.31 + 0.02392 * mean;
        }
    }

    /// <summary>
    /// Actual span (in days) of data within the selected stats time range.
    /// </summary>
    public double StatsDataSpanDays
    {
        get
        {
            var cutoff = DateTime.UtcNow - Settings.StatsTimeRange.Interval();
            var window = RecentReadings.Where(r => r.Date >= cutoff).ToList();
            if (window.Count == 0) return 0;

            var oldest = window.Min(r => r.Date);
            var newest = window.Max(r => r.Date);
            return (newest - oldest).TotalDays;
        }
    }

    /// <summary>
    /// Readings filtered to the selected chart time range, newest first.
    /// </summary>
    public List<GlucoseReading> ChartReadings
    {
        get
        {
            var cutoff = DateTime.UtcNow - Settings.SelectedTimeRange.Interval();
            return RecentReadings.Where(r => r.Date >= cutoff).ToList();
        }
    }

    // -------------------------------------------------------------------------
    // Constructor
    // -------------------------------------------------------------------------

    public GlucoseMonitor()
    {
        _syncContext = SynchronizationContext.Current
            ?? throw new InvalidOperationException(
                "GlucoseMonitor must be constructed on the UI thread.");

        Settings = AppSettings.Load();
        LoadReadingsFromDisk();
    }

    // -------------------------------------------------------------------------
    // Lifecycle
    // -------------------------------------------------------------------------

    /// <summary>
    /// Authenticates with Dexcom and begins polling.
    /// </summary>
    public async Task StartAsync(string username, string password, DexcomRegion region)
    {
        _username = username;
        Settings.Region = region;
        Settings.DexcomUsername = username;

        _service = new DexcomService(region);

        await AuthenticateWithRetryAsync(username, password);

        // Fire an initial poll immediately (288 readings), then switch to incremental.
        _ = Task.Run(async () =>
        {
            await PollAsync(isInitial: true);
            ScheduleTimerFromLastReading();
        });
    }

    /// <summary>
    /// Stops polling, clears session, and sets the monitor to Idle.
    /// </summary>
    public async Task StopAsync()
    {
        await DisposeTimerAsync();

        _service?.ClearSession();
        _service = null;
        _username = null;

        PostToUi(() =>
        {
            State = new MonitorState.Idle();
            NotifyUpdate();
        });
    }

    /// <summary>
    /// Triggers an immediate refresh outside of the regular schedule.
    /// </summary>
    public async Task RefreshNowAsync()
    {
        await DisposeTimerAsync();
        await PollAsync(isInitial: false);
        ScheduleTimerFromLastReading();
    }

    /// <summary>
    /// Updates the polling interval and reschedules the timer immediately.
    /// </summary>
    public void UpdateRefreshInterval(TimeSpan interval)
    {
        Settings.RefreshInterval = interval;
        ScheduleTimerFromLastReading();
    }

    // -------------------------------------------------------------------------
    // Timer scheduling
    // -------------------------------------------------------------------------

    private void ScheduleTimer(TimeSpan dueTime)
    {
        NextRefreshDate = DateTime.UtcNow + dueTime;
        _timer?.Dispose();
        _timer = new System.Threading.Timer(
            callback: async _ => await TimerTickAsync(),
            state: null,
            dueTime: dueTime,
            period: Timeout.InfiniteTimeSpan);
    }

    private void ScheduleTimerFromLastReading()
    {
        if (CurrentReading is null)
        {
            ScheduleTimer(Settings.RefreshInterval);
            return;
        }

        // Align next poll to reading timestamp + refresh interval
        var nextExpected = CurrentReading.Date + Settings.RefreshInterval;
        var dueTime = nextExpected - DateTime.UtcNow;

        if (dueTime < TimeSpan.Zero)
            dueTime = TimeSpan.Zero;

        ScheduleTimer(dueTime);
    }

    private async Task TimerTickAsync()
    {
        await PollAsync(isInitial: false);
        ScheduleTimerFromLastReading();
    }

    private Task DisposeTimerAsync()
    {
        var t = Interlocked.Exchange(ref _timer, null);
        t?.Dispose();
        return Task.CompletedTask;
    }

    // -------------------------------------------------------------------------
    // Polling
    // -------------------------------------------------------------------------

    private async Task PollAsync(bool isInitial)
    {
        if (_service is null) return;

        // Prevent concurrent polls
        if (!await _pollLock.WaitAsync(0)) return;
        try
        {
            PostToUi(() =>
            {
                State = new MonitorState.Loading();
                NotifyUpdate();
            });

            var count = isInitial ? InitialFetchCount : SubsequentFetchCount;
            List<GlucoseReading> fetched;

            try
            {
                fetched = await _service.GetLatestReadingsAsync(count);
            }
            catch (DexcomException ex) when (
                ex.ErrorType is DexcomErrorType.InvalidCredentials or DexcomErrorType.SessionExpired)
            {
                // Session expired — try to re-authenticate
                if (_username is null)
                {
                    PostToUi(() =>
                    {
                        State = new MonitorState.Error("Session expired and no username to re-authenticate.");
                        NotifyUpdate();
                    });
                    return;
                }

                var password = CredentialStorage.LoadPassword();
                if (password is null)
                {
                    PostToUi(() =>
                    {
                        State = new MonitorState.Error("Session expired. Please re-enter credentials.");
                        NotifyUpdate();
                    });
                    return;
                }

                try
                {
                    await AuthenticateWithRetryAsync(_username, password);
                    fetched = await _service.GetLatestReadingsAsync(count);
                }
                catch (Exception retryEx)
                {
                    PostToUi(() =>
                    {
                        State = new MonitorState.Error(retryEx.Message);
                        NotifyUpdate();
                    });
                    return;
                }
            }
            catch (Exception ex)
            {
                PostToUi(() =>
                {
                    State = new MonitorState.Error(ex.Message);
                    NotifyUpdate();
                });
                return;
            }

            MergeReadings(fetched);
            SaveReadingsToDisk();

            var latest = RecentReadings.Count > 0 ? RecentReadings[0] : null;

            PostToUi(() =>
            {
                CurrentReading = latest;
                State = new MonitorState.Connected();
                LastUpdated = DateTime.UtcNow;
                NotifyUpdate();

                if (latest is not null)
                    EvaluateAlerts(latest);
            });
        }
        finally
        {
            _pollLock.Release();
        }
    }

    // -------------------------------------------------------------------------
    // Authentication with retry
    // -------------------------------------------------------------------------

    private async Task AuthenticateWithRetryAsync(string username, string password)
    {
        Exception? lastEx = null;

        for (int attempt = 0; attempt < MaxAuthRetries; attempt++)
        {
            if (attempt > 0)
                await Task.Delay(RetryDelaysMs[attempt - 1]);

            try
            {
                await _service!.AuthenticateAsync(username, password);
                return; // success
            }
            catch (DexcomException ex) when (ex.ErrorType == DexcomErrorType.InvalidCredentials)
            {
                // Invalid credentials won't be fixed by retrying
                throw;
            }
            catch (Exception ex)
            {
                lastEx = ex;
            }
        }

        throw lastEx ?? new DexcomException(DexcomErrorType.Unknown, "Authentication failed.");
    }

    // -------------------------------------------------------------------------
    // Reading merge
    // -------------------------------------------------------------------------

    private void MergeReadings(IEnumerable<GlucoseReading> incoming)
    {
        var existing = new HashSet<DateTime>(RecentReadings.Select(r => r.Date));

        foreach (var r in incoming)
        {
            if (existing.Add(r.Date))
                RecentReadings.Add(r);
        }

        // Sort newest first and cap
        RecentReadings = RecentReadings
            .OrderByDescending(r => r.Date)
            .Take(MaxReadings)
            .ToList();
    }

    // -------------------------------------------------------------------------
    // Alerts
    // -------------------------------------------------------------------------

    private void EvaluateAlerts(GlucoseReading reading)
    {
        // Level alerts — mutually exclusive, checked in priority order
        if (reading.Value <= Settings.AlertUrgentLowThresholdMgdL && Settings.AlertUrgentLowEnabled)
        {
            TryFireAlert("UrgentLow",
                "Urgent Low Alert",
                $"Glucose is {reading.DisplayValue(Settings.Unit)} {reading.Trend.Arrow()} (Urgent Low)");
        }
        else if (reading.Value <= Settings.AlertLowThresholdMgdL && Settings.AlertLowEnabled)
        {
            TryFireAlert("Low",
                "Low Alert",
                $"Glucose is {reading.DisplayValue(Settings.Unit)} {reading.Trend.Arrow()} (Low)");
        }
        else if (reading.Value >= Settings.AlertUrgentHighThresholdMgdL && Settings.AlertUrgentHighEnabled)
        {
            TryFireAlert("UrgentHigh",
                "Urgent High Alert",
                $"Glucose is {reading.DisplayValue(Settings.Unit)} {reading.Trend.Arrow()} (Urgent High)");
        }
        else if (reading.Value >= Settings.AlertHighThresholdMgdL && Settings.AlertHighEnabled)
        {
            TryFireAlert("High",
                "High Alert",
                $"Glucose is {reading.DisplayValue(Settings.Unit)} {reading.Trend.Arrow()} (High)");
        }

        // Trend alerts — independent of level alerts
        if (reading.Trend.IsRisingFast() && Settings.AlertRisingFastEnabled)
        {
            TryFireAlert("RisingFast",
                "Rising Fast",
                "Glucose is rising quickly");
        }

        if (reading.Trend.IsDroppingFast() && Settings.AlertDroppingFastEnabled)
        {
            TryFireAlert("DroppingFast",
                "Dropping Fast",
                "Glucose is dropping quickly");
        }

        // Stale data alert
        if (IsStale && Settings.AlertStaleDataEnabled)
        {
            TryFireAlert("StaleData",
                "Stale Data",
                "No readings in 20 minutes");
        }
    }

    private void TryFireAlert(string key, string title, string message)
    {
        var now = DateTime.UtcNow;

        if (_alertCooldowns.TryGetValue(key, out var lastFired) &&
            now - lastFired < AlertCooldown)
        {
            return; // still in cooldown
        }

        _alertCooldowns[key] = now;
        OnAlert?.Invoke(title, message);
    }

    // -------------------------------------------------------------------------
    // Disk persistence
    // -------------------------------------------------------------------------

    private void LoadReadingsFromDisk()
    {
        try
        {
            if (!File.Exists(ReadingsPath))
                return;

            var json = File.ReadAllText(ReadingsPath);
            var loaded = JsonSerializer.Deserialize<List<PersistedReading>>(json, JsonOptions);
            if (loaded is null) return;

            RecentReadings = loaded
                .Select(p => new GlucoseReading
                {
                    Id = p.Id,
                    Value = p.Value,
                    Trend = p.Trend,
                    Date = p.Date,
                    TrendRate = p.TrendRate
                })
                .OrderByDescending(r => r.Date)
                .Take(MaxReadings)
                .ToList();

            CurrentReading = RecentReadings.Count > 0 ? RecentReadings[0] : null;
        }
        catch
        {
            // Non-fatal: start with empty readings
            RecentReadings = [];
        }
    }

    private void SaveReadingsToDisk()
    {
        try
        {
            var dir = Path.GetDirectoryName(ReadingsPath)!;
            Directory.CreateDirectory(dir);

            var persisted = RecentReadings.Select(r => new PersistedReading
            {
                Id = r.Id,
                Value = r.Value,
                Trend = r.Trend,
                Date = r.Date,
                TrendRate = r.TrendRate
            }).ToList();

            var json = JsonSerializer.Serialize(persisted, JsonOptions);
            File.WriteAllText(ReadingsPath, json);
        }
        catch
        {
            // Non-fatal: disk save failure should not crash the app
        }
    }

    // -------------------------------------------------------------------------
    // Helpers
    // -------------------------------------------------------------------------

    /// <summary>Posts an action back to the captured UI synchronization context.</summary>
    private void PostToUi(Action action) =>
        _syncContext.Post(_ => action(), null);

    /// <summary>Fires OnUpdate. Must be called from the UI thread (inside PostToUi).</summary>
    private void NotifyUpdate() => OnUpdate?.Invoke();

    // -------------------------------------------------------------------------
    // IDisposable
    // -------------------------------------------------------------------------

    public void Dispose()
    {
        if (_disposed) return;
        _disposed = true;

        _timer?.Dispose();
        _timer = null;
        _pollLock.Dispose();
        _service?.ClearSession();
        GC.SuppressFinalize(this);
    }

    // -------------------------------------------------------------------------
    // Private DTO for JSON persistence (avoids init-only property issues)
    // -------------------------------------------------------------------------

    private sealed class PersistedReading
    {
        public Guid Id { get; set; }
        public int Value { get; set; }
        public GlucoseTrend Trend { get; set; }
        public DateTime Date { get; set; }
        public double? TrendRate { get; set; }
    }
}
